CMakeLists.txt

```diff
@@ -62,22 +74,30 @@ endif (CMAKE_INTERPROCEDURAL_OPTIMIZATION)
 set(C_EXTRA_GNU )
 set(C_EXTRA_Clang )
 
+if (ENABLE_PROFILE_GUIDED_OPTIMIZATION)
+    message(STATUS "Profile guided optimization is enabled on ${CMAKE_C_COMPILER_ID} compiler")
+    # Set the build type to coverage since we need coverage related
+    # build options when -fprofile-generate is used.
+    if("GNU" STREQUAL ${CMAKE_C_COMPILER_ID})
+        # Temporarily modify the CMAKE_BUILD_TYPE to Coverage. Note that we are not setting the
+        # CMAKE_BUILD_TYPE to Coverage using the CACHE option. This means that when this build is over,
+        # we will retain the original value of CMAKE_BUILD_TYPE
+        set(CMAKE_BUILD_TYPE Coverage)
+        #  You must use -fprofile-generate both when compiling *and* when linking your program
+        set(C_EXTRA_GNU ${C_EXTRA_GNU} -fprofile-generate=${CMAKE_BINARY_DIR}/profile-data -fprofile-correction)
```

Comment:
I happened to run across https://go.googlesource.com/proposal/+/master/design/55022-pgo.md, which discussed two different methods of doing PGO. Either tracing coverage (what this PR does), or a newer approach of sampling coverage in production, which is what the linked document also explains in detail.

See the heading "AutoFDO vs. instrumentation-based FDO".

My conclusion from reading the Go proposal is that for skupper-router, AutoFDO is not necessary and that `-fprofile-generate` is good enough.

Interesting quote:

> Google‘s AutoFDO (for C/C++ programs) is built on LLVM’s sample-based PGO, along with other toolings and mechanisms such as [Google-wide profiling (GWP)](https://research.google/pubs/pub36575). [AutoFDO](https://research.google/pubs/pub45290) improves the performance of C/C++ programs by 5–15% in Google datacenters.